### PR TITLE
VTN-9746 - Retry on 502 and provide client DebugLog with insight in request & response

### PR DESCRIPTION
**Pull request by dduyVeritone**
## Overview
* Retry on 502
* Add debug log function option for client

## Testing
* Unit test passed
* Log looks like
```
	graphql_retry_test.go:120: Request: &{Method:POST URL:http://127.0.0.1:55503 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Content-Type:[application/json; charset=utf-8] Accept:[application/json; charset=utf-8]] Body:{Reader:} GetBody:0x1273dc0 ContentLength:38 TransferEncoding:[] Close:false Host:127.0.0.1:55503 Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr: RequestURI: TLS:<nil> Cancel:<nil> Response:<nil> ctx:0xc420071320}
	graphql_retry_test.go:121: Response: &{Status:502 Bad Gateway StatusCode:502 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Content-Length:[0] Date:[Thu, 16 Aug 2018 20:44:43 GMT]] Body:{} ContentLength:0 TransferEncoding:[] Close:false Uncompressed:false Trailer:map[] Request:0xc42012ab00 TLS:<nil>}
```
with
```go
	client.DebugLog = func(req *http.Request, resp *http.Response) {
		t.Logf("Request: %+v", req)
		t.Logf("Response: %+v", resp)
	}
```

### CODE REVIEW

File: graphql.go [hunk 1]
```diff
@@ -55,13 +55,16 @@ type Client struct {
 	// To log to standard out, use:
 	//  client.Log = func(s string) { log.Println(s) }
 	Log func(s string)
+	// Client can use this function to supply some logging logic to further debug GraphQL request & response
+	DebugLog func(req *http.Request, resp *http.Response)
```
Comment (mezzi) [inline graphql.go]: Can you name this something like `OnRetry` since it is invoked on retries and can have other logic than just logging?

Comment (mezzi) [reply graphql.go]: or `BeforeRetry` might be more appropriate

Comment (dduyVeritone) [reply graphql.go]: Ok will do.

File: graphql.go [hunk 2]
```diff
@@ -138,6 +141,7 @@ func (c *Client) sendRequest(req *http.Request) (*http.Response, error) {
 		if err == nil && !retryConfig.shouldRetry(resp.StatusCode) {
 			return resp, nil
 		}
+		c.BeforeRetry(req, resp)
```
Comment (mezzi) [inline graphql.go]: add a check for `c.BeforeRetry != nil`

**Pull request merged**